### PR TITLE
Implemented dot() and cross() methods

### PR DESCRIPTION
**Pull request by ShikharJ**
Relevant: https://github.com/symengine/symengine.py/issues/17
Currently implementing `norm` is out of scope as for that we need to have a method for finding eigenvalues, which is planned to be implemented as a part of Ranjith's project.

### CODE REVIEW

File: symengine/dense_matrix.cpp [hunk 1]
```diff
@@ -1520,6 +1520,57 @@ void inverse_gauss_jordan(const DenseMatrix &A, DenseMatrix &B)
...
+        for (unsigned j = 0; j < A.col_; j++) {
+            tmp1.m_[i * A.col_ + j] = A.m_[i * A.col_ + j];
+        }
+    }
+
+    for (unsigned i = 0; i < B.row_; i++) {
+        for (unsigned j = 0; j < B.col_; j++) {
+            tmp2.m_[i * B.col_ + j] = B.m_[i * B.col_ + j];
+        }
+    }
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Can you remove the above unnecessary copying and use them only if they are needed?

**Comment by ShikharJ** [comment]
Ping @isuruf

File: symengine/dense_matrix.cpp [hunk 2]
```diff
@@ -1520,6 +1520,66 @@ void inverse_gauss_jordan(const DenseMatrix &A, DenseMatrix &B)
     fraction_free_gauss_jordan_solve(A, e, B);
 }
 
+// ----------------------- Vector-specific Methods --------------------------//
+
+void dot(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+{
+    DenseMatrix tmp1 = DenseMatrix(A.row_, A.col_);
+    DenseMatrix tmp2 = DenseMatrix(B.row_, B.col_);
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: This still allocates unnecessarily.

File: symengine/dense_matrix.cpp [hunk 3]
```diff
@@ -1520,6 +1520,66 @@ void inverse_gauss_jordan(const DenseMatrix &A, DenseMatrix &B)
...
+            for (unsigned i = 0; i < A.row_; i++) {
+                for (unsigned j = 0; j < A.col_; j++) {
+                    tmp1.m_[i * A.col_ + j] = A.m_[i * A.col_ + j];
+                }
+            }
+            for (unsigned i = 0; i < B.row_; i++) {
+                for (unsigned j = 0; j < B.col_; j++) {
+                    tmp2.m_[i * B.col_ + j] = B.m_[i * B.col_ + j];
+                }
+            }
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: This copy is redundant, you can do
```
C.resize(A.row_, B.col_);
mul_dense_dense(A, B, C);
```

File: symengine/dense_matrix.cpp [hunk 4]
```diff
@@ -1520,6 +1520,66 @@ void inverse_gauss_jordan(const DenseMatrix &A, DenseMatrix &B)
...
+        dot(tmp1, tmp2, C);
+    } else if (tmp1.row_ == tmp2.row_) {
+        for (unsigned i = 0; i < B.row_; i++) {
+            for (unsigned j = 0; j < B.col_; j++) {
+                tmp2.m_[i * B.col_ + j] = B.m_[i * B.col_ + j];
+            }
+        }
+        tmp1.resize(A.col_, A.row_);
+        A.transpose(tmp1);
+        dot(tmp1, tmp2, C);
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Change above 9 lines to
```
tmp1.resize(A.col_, A.row_);
A.transpose(tmp1);
dot(tmp1, B, C);
```

**Comment by ShikharJ** [comment]
Ping @isuruf

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/dense_matrix.cpp [hunk 5]
```diff
@@ -1520,6 +1520,45 @@ void inverse_gauss_jordan(const DenseMatrix &A, DenseMatrix &B)
...
+            mul_dense_dense(A, B, C);
+        }
+    } else if (A.col_ == B.col_) {
+        DenseMatrix tmp2 = DenseMatrix(B.col_, B.row_);
+        B.transpose(tmp2);
+        dot(A, tmp2, C);
+    } else if (A.row_ == B.row_) {
+        DenseMatrix tmp1 = DenseMatrix(A.col_, A.row_);
+        A.transpose(tmp1);
+        dot(tmp1, B, C);
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Can you add a test for this as well?

Comment (ShikharJ) [inline symengine/dense_matrix.cpp]: I think that this case is redundant, even though `SymPy` has  it implemented. I'll remove it.

**Comment by ShikharJ** [comment]
This can be merged as it is. All the builds complete and the tests pass. The only failure is due to some new issue while running the `codecov` script, after the tests.

**Comment by ShikharJ** [comment]
Ping @isuruf

**Comment by certik** [comment]
Thanks @ShikharJ, that looks good.

**Pull request merged**